Salt-marsh-metasecretome-analysis
Salt marsh meta-exo-proteome analysis

![Saltmarsh layout](https://github.com/leadbot/Salt-marsh-metasecretome-analysis/blob/master/Images/Cage%2520tide.jpg)

Here you will find consolidated scripts, package dependencies and inputs required to replicate our analyses

For the quickest, cleanest and easiest access to the consolidated data, please look at the **Consolidated_data_file.xlsx** which contains Molar percentages, dbCAN annotations and Taxonomic homology for all CAZy proteins - the rest of these folders contain scripts to perform non filtered analysis (>16Gb RAM required) 


Saltmarsh_metasecretome_python_environment.yml contains the python environment and dependencies required to run these scripts

Biomass composition analysis contains all raw data and scripts required to analyse the data and generate the figures

**ISDE_Master_mascot_COMPOSITE_100k_search_0pt05DB_with_0pt1sig_11268U.xlsx** contains all proteomic data from the secondary 'true hit' peptide search

**dbCAN_OUTPUT_DB0pt05_vs_0pt1SIG.xlsx** contains all dbCAN CAZy domain annotations


![Cage deployment](https://github.com/leadbot/Salt-marsh-metasecretome-analysis/blob/master/Images/Figure3.png)


![Cage deployment](https://github.com/leadbot/Salt-marsh-metasecretome-analysis/blob/master/Images/Schematic%2520diag%2520-%2520NB%26DL%2520-%2520230719-01.jpg)
